script/src/verify.rs

```diff
@@ -98,12 +124,37 @@ impl<'a, CS: ChainStore> TransactionScriptsVerifier<'a, CS> {
             .collect();
 
         let mut block_data = FnvHashMap::<H256, (BlockNumber, BlockExt)>::default();
-        for resolved_input in &resolved_inputs {
+        let mut lock_groups = FnvHashMap::default();
+        let mut type_groups = FnvHashMap::default();
+        for (i, resolved_input) in resolved_inputs.iter().enumerate() {
             if let Some(header) = &resolved_input.header {
                 if let Some(block_ext) = store.get_block_ext(header.hash()) {
                     block_data.insert(header.hash().to_owned(), (header.number(), block_ext));
                 }
             }
+            // here we are only pre-processing the data, verify method validates
+            // each input has correct script setup.
+            if let Some(cell_meta) = resolved_input.cell.cell_meta() {
+                let output = store.lazy_load_cell_output(cell_meta);
+                let lock_group_entry = lock_groups
+                    .entry(output.lock.hash())
+                    .or_insert_with(|| ScriptGroup::new(&output.lock));
+                lock_group_entry.input_indices.push(i);
+                if let Some(t) = output.type_ {
```

Comment:
Why do we need to care about the type of previous output?

Reply:
In the new verification model, type scripts from both inputs and outputs will be run when validating a transaction.